src/biogeochem/SPMMod.F90

```diff
@@ -0,0 +1,805 @@
+         i_ab = i_ab + 1
+      else
+         if(Aindex(i_a) .gt. Bindex(i_b))then 
+            do fu = 1, num_unit
+               u = filter_u(fu)
+               this%M(u,i_ab) = B%M(u,i_b)
+            end do
+            ABindex(i_ab) = Bindex(i_b)
+            if(present(list_B))list_B(i_b) = i_ab
+            i_b  = i_b  + 1
+            i_ab = i_ab + 1
+         else
+            do fu = 1, num_unit
+               u = filter_u(fu)
+               this%M(u,i_ab) = A%M(u,i_a) + B%M(u,i_b)
+            end do
+            ABindex(i_ab) = Aindex(i_a)
+            if(present(list_A))list_A(i_a) = i_ab
+            if(present(list_B))list_B(i_b) = i_ab
+            i_a  = i_a  + 1
+            i_b  = i_b  + 1
+            i_ab = i_ab + 1
+         end if
+      end if
+   end do
+
+   this%NE = i_ab - 1
+   this%CI(1:this%NE) = (ABindex(1:this%NE) - 1) / this%SM + 1
+   this%RI(1:this%NE) =  ABindex(1:this%NE) - this%SM * (this%CI(1:this%NE) - 1)
+   if(present(NE_AB))NE_AB = this%NE
+   if(present(CI_AB))CI_AB(1:this%NE) = this%CI(1:this%NE)
+   if(present(RI_AB))RI_AB(1:this%NE) = this%RI(1:this%NE)
+   if(present(list_A) .and. present(list_B) .and. present(NE_AB) .and. present(RI_AB) .and. present(CI_AB))list_ready = .true.
+else
+   do i = 1, NE_AB
+      do fu = 1, num_unit
+         u = filter_u(fu)
+         this%M(u,i) = 0._r8
+      end do
+   end do
+   do i_a = 1, A%NE
+      do fu = 1, num_unit
+         u = filter_u(fu)
+         this%M(u,list_A(i_a)) = A%M(u,i_a)
+      end do
+   end do
+   do i_b = 1, B%NE
+      do fu = 1, num_unit
+         u = filter_u(fu)
+         this%M(u,list_B(i_b)) = this%M(u,list_B(i_b)) + B%M(u,i_b)
+      end do
+   end do
+   this%NE = NE_AB
+   this%CI(1:this%NE) = CI_AB(1:NE_AB)
+   this%RI(1:this%NE) = RI_AB(1:NE_AB)
+end if
+
+end subroutine SPMP_AB
+
+subroutine SPMP_ABC(this,num_unit,filter_u,A,B,C,list_ready,list_A,list_B,list_C,NE_ABC,RI_ABC,CI_ABC,&
+           use_actunit_list_A,num_actunit_A,filter_actunit_A,use_actunit_list_B,num_actunit_B,filter_actunit_B,&
+           use_actunit_list_C,num_actunit_C,filter_actunit_C)
+
+! Calculate sparse matrix addition (SPMP) ABC(this) =  A + B + C
+! The map of each entry in A, B and C to ABC have been memorized to save the computational cost, 
+! since they are usually time-independent. 
+! The information of active landunit/gridcell/column/patch is used to save computational cost.
+! A is a sparse matrix in Coordinate format (COO)
+! B is a sparse matrix in Coordinate format (COO)
+! C is a sparse matrix in Coordinate format (COO)
+! ABC is a sparse matrix in Coordinate format (COO)
+
+class(sparse_matrix_type)     :: this
+type(sparse_matrix_type),intent(in)     :: A
+type(sparse_matrix_type),intent(in)     :: B
+type(sparse_matrix_type),intent(in)     :: C
+logical,intent(inout)                   :: list_ready 
+integer,intent(in) :: num_unit
+integer,intent(in) :: filter_u(:)
+logical,intent(in),optional :: use_actunit_list_A
```

Comment:
@chrislxj there's a lot of additional complexity and code duplication to support this functionality of having different filters for each matrix. As far as I can see it's not being used. Is this something needed that I missed a reference to? Or is it something needed for the future? If not we should remove it.

Reply:
We should re-examine this question about if we really need different filters for each matrix in SPMP_ABC. I asked the question and never got back to it.